pkg/cached/cache.go

```diff
@@ -286,23 +292,21 @@ type Replaceable[T any] struct {
 // previously had returned a success, that success will be returned
 // instead. If the cache fails but we never returned a success, that
 // failure is returned.
-//
-// # Important note
-//
-// As all implementations of Get, this implementation is NOT
-// thread-safe. Please properly lock a mutex before calling this method
-// if you are in a multi-threaded environment, since this method will
-// update the cache and perform write operations.
 func (c *Replaceable[T]) Get() Result[T] {
 	result := (*c.cache.Load()).Get()
-	if result.Err != nil && c.result != nil && c.result.Err == nil {
-		return *c.result
+
+	for {
```

Comment:
It's cool that you did this without a mutex, the logic feels a bit complex and I'm not sure if we get much performance benefit. Logically it makes sense that we'll terminate when the CompareAndSwap succeeds but having an infinite for loop could cause problems if we're not careful. wdyt?

Reply:
Yeah, I don't think the infinite loop is so bad but I might be wrong, feel free to think of a test that could trigger that?
Complexity is a little higher than mutexes, my main goal here is that I don't want to lock both in Get and in Replace. I could still use a mutex in Get and keep the atomic pointer for replace.

Reply:
>  I could still use a mutex in Get and keep the atomic pointer for replace.
I would probably slightly prefer that approach but will defer the decision to you. 

This might already be a pattern we use in aggregated discovery so I don't feel too strongly but did want to bring it up. Only case I can think of is a thread being starved and constantly losing the compareandswap operation which would increase latency on the get. However, this case probably never happens and we have bigger problems if we have such a high frequency of reads/writes to the cache 😄